### PR TITLE
Refactor Result class

### PR DESCRIPTION
**Pull request by adasiewiczr**
### 🎯 Goal

Refactor `Result` class.

### 🛠 Implementation details

1. Changed new `Result` class definition - replaced unnecessary generic parameter with `ChatError`
2. Removed old `Result` implementation
3. Migrated codebase
4. Migrated snippets

### 🧪 Testing

Do regression testing on sample apps.

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [x] Changelog is updated with client-facing changes
- [x] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### CODE REVIEW

File: stream-chat-android-core/src/main/java/io/getstream/chat/android/client/utils/Result.kt [hunk 1]
```diff
@@ -17,221 +17,129 @@
...
-
-    @Suppress("DEPRECATION")
-    public constructor(data: T) : this(data, null)
-
-    @Suppress("DEPRECATION")
-    public constructor(error: ChatError) : this(null, error)
+public sealed class Result<out A : Any> {
 
     /**
-     * Returns true if a request of payload response has been successful.
```
Comment (skydoves) [inline stream-chat-android-core/src/main/java/io/getstream/chat/android/client/utils/Result.kt:37]: What do you think adding some extensions for Result class below?

```kotlin
public val Result<*>.isSuccess: Boolean
    inline get() = this is Result.Success

public val Result<*>.isFailure: Boolean
    inline get() = this is Result.Failure
```

There're many useful cases for this and we also don't need to change Java codes to check `instanceOf`.

**Review by skydoves — COMMENTED** [review]
Overall looks good to me!
I left a trivial comment :)

File: stream-chat-android-core/src/main/java/io/getstream/chat/android/client/utils/Result.kt (hunk 1 above)
Comment (adasiewiczr) [reply stream-chat-android-core/src/main/java/io/getstream/chat/android/client/utils/Result.kt:37]: Good idea! Let me update the code

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/notifications/LoadNotificationDataWorker.kt [hunk 2]
```diff
@@ -61,13 +61,16 @@ internal class LoadNotificationDataWorker(
...
             val result = getChannel.zipWith(getMessage).await()
-            if (result.isSuccess) {
-                val (channel, message) = result.data()
-                ChatClient.displayNotification(channel = channel, message = message)
-                Result.success()
-            } else {
-                logger.e { "Error while loading notification data: ${result.error().stringify()}" }
-                Result.failure()
+            when (result) {
+                is io.getstream.chat.android.client.utils.Result.Success -> {
```
Comment (leandroBorgesFerreira) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/notifications/LoadNotificationDataWorker.kt:65]: I think we don't need the whole import path here.

**Review by leandroBorgesFerreira — APPROVED** [review]

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/StreamFileUtil.kt [hunk 3]
```diff
@@ -127,45 +127,47 @@ public object StreamFileUtil {
...
-            if (isFileCached) {
-                Result(data = getUriForFile(context, file))
-            } else {
-                Result(
-                    error = ChatError(
-                        message = "No such file in cache.",
-                    )
-                )
+            when (val getOrCreateCacheDirResult = getOrCreateStreamCacheDir(context)) {
+                is Result.Failure -> return getOrCreateCacheDirResult
```
Comment (leandroBorgesFerreira) [inline stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/StreamFileUtil.kt:141]: This `return` is not needed.

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/StreamFileUtil.kt [hunk 4]
```diff
@@ -192,47 +194,48 @@ public object StreamFileUtil {
...
-
-                val response = ChatClient.instance().downloadFile(fileUrl).await()
-
-                if (response.isSuccess) {
-                    // write the response to a file
-                    response.data().byteStream().use { inputStream ->
-                        file.outputStream().use { outputStream ->
-                            inputStream.copyTo(outputStream)
+            when (val getOrCreateCacheDirResult = getOrCreateStreamCacheDir(context)) {
+                is Result.Failure -> return getOrCreateCacheDirResult
```
Comment (leandroBorgesFerreira) [inline stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/StreamFileUtil.kt:198]: This return is not needed

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/StreamFileUtil.kt [hunk 5]
```diff
@@ -192,47 +194,48 @@ public object StreamFileUtil {
...
+                is Result.Success -> {
+                    val streamCacheDir = getOrCreateCacheDirResult.value
+
+                    val attachmentHashCode = (attachment.url ?: attachment.assetUrl)?.hashCode()
+                    val fileName = CACHED_FILE_PREFIX + attachmentHashCode.toString() + attachment.name
+
+                    val file = File(streamCacheDir, fileName)
+
+                    return if (file.exists() &&
```
Comment (leandroBorgesFerreira) [inline stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/StreamFileUtil.kt:207]: This return is not needed.

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/notifications/LoadNotificationDataWorker.kt (hunk 2 above)
Comment (adasiewiczr) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/notifications/LoadNotificationDataWorker.kt:65]: Unfortunately we need it as below we are returning Result from `androidx.work` package

**Pull request merged**